sunxspex/sunxspex_fitting/fitter.py

```diff
@@ -0,0 +1,4931 @@
+    environment to the one the original function was defined in.
+
+    If an exception occurs here then the user is informed with a warning and
+    (hopefully) a helpful message.
+
+    The test inputs to check a function works are 1s for all arguments and
+    energy bins of np.array([[1.60,1.64],[1.64,1.68],...,[4.96,5.00]]).
+
+    Parameters
+    ----------
+    function_name : str
+            The name of the function.
+
+    function_text : str
+            Code for the function as a string.
+
+    Returns
+    -------
+    None.
+    """
+    exec(function_text, globals())
+    params, _ = get_func_inputs(globals()[function_name])
+    _test_e_range = np.arange(1.6,5.01, 0.04)[:,None]
+    _test_params, _test_energies = np.ones(len(params)), np.concatenate((_test_e_range[:-1], _test_e_range[1:]), axis=1) # one 5 for each param, 2 column array of e-bins
+    try:
+        _func_to_test = globals()[function_name]
+        del globals()[function_name] # this is a check function, don't want it just adding things to globals
+        _func_to_test(*_test_params, energies=_test_energies)
+    except NameError as e:
+        raise NameError(str(e)+f"\nA user defined function should be completely self-contained and should be able to be created from its source code,\nrelying entirely on its local scope. E.g., modules not imported in the fitter module need to be imported in the fuction (fitter imported modules:\n{imports()}).")
+    except FileNotFoundError as e:
+        raise FileNotFoundError(str(e)+"\nPlease check that absolute file/directory paths are given for the user defined function to be completely self-contained.")
+    # except ValueError as e:
+    #     warnings.warn(str(e)+"\nFunction failed self-contained check; however, this may be due to conflict in test inputs used to the model.")
+
+def function_creator(function_name, function_text, _orig_func=None):
+    """ Creates a named function from its name and source code.
+
+    Takes a user defined function name for a NAMED function and a string that
+    produces the NAMED function and executes the string as code. Replicates the
+    user coding thier function in this module directly.
+
+    Parameters
+    ----------
+    function_name : str
+            The name of the function.
+
+    function_text : str
+            Code for the function as a string.
+
+    _orig_func : function or None
+            The original function provided to be broken down and recreated.
+            Default: None
+
+    Returns
+    -------
+    Returns the function that has just been created and executed into globals().
+    """
+    DYNAMIC_FUNCTION_SOURCE[function_name] = function_text
+    try:
+        _func_self_contained_check(function_name, function_text)
+        # given the code for a NAMED function (not lambda) as a string, this will execute the code and return that function
+        exec(function_text, globals())
```

Comment:
Absolutely agree. I think I have discussed this in another comment but this was a solution to a pickling issue. I think I have since come up with a better solution by far, although I would love to discuss other methods of doing this.